### PR TITLE
[feature-wip](workload-group) Support for workload group Authentication

### PR DESCRIPTION
**Pull request by luozenglin**
## Proposed changes

Issue Number: close #xxx

1. Support for grant and revoke workload group usage_priv to users or roles:
```
GRANT USAGE_PRIV ON WORKLOAD GROUP 'g1' TO user1;
REVOKE USAGE_PRIV ON WORKLOAD GROUP 'g1' FROM user1;
```
2. Authorized workload group names support the `%` match character, but not `*`.
3. workload_groups() tvf function can view all workload groups, does not support authentication for now.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java [hunk 1]
```diff
@@ -55,10 +56,12 @@ public class Role implements Writable, GsonPostProcessable {
...
     public static Role OPERATOR = new Role(OPERATOR_ROLE,
             TablePattern.ALL, PrivBitSet.of(Privilege.NODE_PRIV, Privilege.ADMIN_PRIV),
-            ResourcePattern.ALL, PrivBitSet.of(Privilege.NODE_PRIV, Privilege.ADMIN_PRIV));
+            ResourcePattern.ALL, PrivBitSet.of(Privilege.NODE_PRIV, Privilege.ADMIN_PRIV), WorkloadGroupPattern.ALL,
+            PrivBitSet.of(Privilege.USAGE_PRIV));
     public static Role ADMIN = new Role(ADMIN_ROLE,
             TablePattern.ALL, PrivBitSet.of(Privilege.ADMIN_PRIV),
-            ResourcePattern.ALL, PrivBitSet.of(Privilege.ADMIN_PRIV));
+            ResourcePattern.ALL, PrivBitSet.of(Privilege.ADMIN_PRIV), WorkloadGroupPattern.ALL,
+            PrivBitSet.of(Privilege.USAGE_PRIV));
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java:64]: Admin only has USAGE_PRIV? not ADMIN_PRIV?

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/WorkloadGroupPrivEntry.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        if (privs.containsNodePriv() || privs.containsDbTablePriv()) {
+            throw new AnalysisException(
+                    "Workload group privilege can not contains node or db table privileges: " + privs);
+        }
+        return new WorkloadGroupPrivEntry(workloadGroupPattern,
+                workloadGroupName, privs);
+    }
+
+    public PatternMatcher getWorkloadGroupPattern() {
+        return workloadGroupPattern;
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/WorkloadGroupPrivEntry.java:63]: it seems ```workloadGroupPatternMatcher``` is better.

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/WorkloadGroupPrivEntry.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            isAnyWorkloadGroup = true;
+        }
+    }
+
+    public static WorkloadGroupPrivEntry create(String workloadGroupName, PrivBitSet privs)
+            throws AnalysisException, PatternMatcherException {
+        PatternMatcher workloadGroupPattern = PatternMatcher.createMysqlPattern(
+                workloadGroupName.equals(ANY_WORKLOAD_GROUP) ? "%" : workloadGroupName,
+                CaseSensibility.WORKLOAD_GROUP.getCaseSensibility());
+        if (privs.containsNodePriv() || privs.containsDbTablePriv()) {
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/WorkloadGroupPrivEntry.java:54]:
```suggestion
        if (privs.containsNodePriv() || privs.containsDbTablePriv() || privs.containsResourcePriv()) {
```

**Comment by luozenglin** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/WorkloadGroupPrivEntry.java (hunk 2 above)
Comment (luozenglin) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/WorkloadGroupPrivEntry.java:63]: This is consistent with other PrivEntry

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java [hunk 4]
```diff
@@ -103,15 +117,24 @@ public Role(String roleName, ResourcePattern resourcePattern, PrivBitSet privs)
...
+            PrivBitSet workloadGroupPrivs) {
         this.roleName = roleName;
         this.tblPatternToPrivs.put(tablePattern, tablePrivs);
         this.resourcePatternToPrivs.put(resourcePattern, resourcePrivs);
+        this.workloadGroupPatternToPrivs.put(workloadGroupPattern, workloadGroupPrivs);
         //for init admin role,will not generate exception
         try {
             grantPrivs(tablePattern, tablePrivs.copy());
             grantPrivs(resourcePattern, resourcePrivs.copy());
+            grantPrivs(workloadGroupPattern, workloadGroupPrivs);
```
Comment (wangbo) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java:137]: Why not copy here? just like tablePrivs and resourcePrivs

**Comment by luozenglin** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/WorkloadGroupMgr.java [hunk 5]
```diff
@@ -133,6 +135,19 @@ public QueryQueue getWorkloadGroupQueryQueue(String groupName) throws UserExcept
         }
     }
 
+    private String getWorkloadGroupName(ConnectContext context) throws AnalysisException {
+        String groupName = context.getSessionVariable().getWorkloadGroup();
+        if (Strings.isNullOrEmpty(groupName)) {
+            groupName = Env.getCurrentEnv().getAuth().getWorkloadGroup(context.getQualifiedUser());
+        }
+        if (!Env.getCurrentEnv().getAccessManager().checkWorkloadGroupPriv(context, groupName, PrivPredicate.USAGE)) {
```
Comment (liutang123) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/WorkloadGroupMgr.java:143]: If `groupName` is still null, set it to null?

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/WorkloadGroupMgr.java [hunk 6]
```diff
@@ -133,6 +135,19 @@ public QueryQueue getWorkloadGroupQueryQueue(String groupName) throws UserExcept
...
+            ErrorReport.reportAnalysisException(
+                    "Access denied; you need (at least one of) the %s privilege(s) to use workload group '%s'.",
```
Comment (liutang123) [inline fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/WorkloadGroupMgr.java:145]: Should we check permissions when `setWorkloadGroup` in session?

Comment (luozenglin) [reply fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/WorkloadGroupMgr.java:145]: already done this

File: fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/WorkloadGroupMgr.java (hunk 5 above)
Comment (luozenglin) [reply fe/fe-core/src/main/java/org/apache/doris/resource/workloadgroup/WorkloadGroupMgr.java:143]: The `default_workload_group` in the user property is `normal` by default

**Comment by luozenglin** [comment]
run buildall

**Review by morningman — CHANGES_REQUESTED** [review]
Please add regression test or unit test for this feature.
(FE ut may be easier than regression test, see AuthTest.java)

**Comment by luozenglin** [comment]
run buildall

File: docs/zh-CN/docs/sql-manual/sql-reference/Account-Management-Statements/GRANT.md [hunk 7]
```diff
@@ -68,7 +70,7 @@ privilege_list 是需要赋予的权限列表，以逗号分隔。当前 Doris 
 
     1. 节点权限：NODE_PRIV
     2. 库表权限：SELECT_PRIV,LOAD_PRIV,ALTER_PRIV,CREATE_PRIV,DROP_PRIV
-    3. 资源权限：USAGE_PRIV
+    3. 资源权限<version since="dev">和workload group权限</version>：USAGE_PRIV
```
[the remark this answers is not in the document]
Comment (luozenglin) [reply docs/zh-CN/docs/sql-manual/sql-reference/Account-Management-Statements/GRANT.md]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/GrantStmt.java [hunk 8]
```diff
@@ -69,6 +71,21 @@ public GrantStmt(UserIdentity userIdent, String role,
         this.role = role;
         this.tblPattern = null;
         this.resourcePattern = resourcePattern;
+        this.workloadGroupPattern = null;
```
[the remark this answers is not in the document]
Comment (luozenglin) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/GrantStmt.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java [hunk 9]
```diff
@@ -53,12 +55,21 @@ public class Role implements Writable, GsonPostProcessable {
...
-            ResourcePattern.ALL, PrivBitSet.of(Privilege.NODE_PRIV, Privilege.ADMIN_PRIV));
-    public static Role ADMIN = new Role(ADMIN_ROLE,
-            TablePattern.ALL, PrivBitSet.of(Privilege.ADMIN_PRIV),
-            ResourcePattern.ALL, PrivBitSet.of(Privilege.ADMIN_PRIV));
+    public static Role OPERATOR;
+    public static Role ADMIN;
+
+    static {
+        try {
+            // Global privileges do not distinguish Patter, global admin and node privileges contain all privileges
```
[the remark this answers is not in the document]
Comment (luozenglin) [reply fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/Role.java]: done

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**